flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecLegacySink.java

```diff
@@ -0,0 +1,182 @@
+                    } else {
+                        throw new TableException(
+                                "UpsertStreamTableSink requires that Table has a full primary keys if it is updated.");
+                    }
+                }
+
+                transform = translateToTransformation(planner, true);
+            } else if (tableSink instanceof AppendStreamTableSink) {
+                // verify table is an insert-only (append-only) table
+                if (needRetraction) {
+                    throw new TableException(
+                            "AppendStreamTableSink requires that Table has only insert changes.");
+                }
+                transform = translateToTransformation(planner, false);
+            } else {
+                throw new TableException(
+                        "Stream Tables can only be emitted by AppendStreamTableSink, "
+                                + "RetractStreamTableSink, or UpsertStreamTableSink.");
+            }
+
+            final DataStream<T> dataStream = new DataStream<T>(planner.getExecEnv(), transform);
+            final DataStreamSink<?> dsSink =
+                    ((StreamTableSink<T>) tableSink).consumeDataStream(dataStream);
+            if (dsSink == null) {
+                throw new TableException(
+                        String.format(
+                                "The StreamTableSink#consumeDataStream(DataStream) must be implemented "
+                                        + "and return the sink transformation DataStreamSink. "
+                                        + "However, %s doesn't implement this method.",
+                                tableSink.getClass().getCanonicalName()));
+            }
+            return dsSink.getTransformation();
+        } else if (tableSink instanceof DataStreamTableSink) {
+            // In case of table to DataStream through
+            // StreamTableEnvironment#toAppendStream/toRetractStream,
+            // we insert a DataStreamTableSink that wraps the given DataStream as a LogicalSink. It
+            // is no real table sink, so we just need translate its input to Transformation.
+            return (Transformation<Object>)
+                    translateToTransformation(
+                            planner, ((DataStreamTableSink<T>) tableSink).withChangeFlag());
+        } else {
+            throw new TableException(
+                    String.format(
+                            "Only Support StreamTableSink! However %s is not a StreamTableSink.",
+                            tableSink.getClass().getCanonicalName()));
+        }
+    }
+
+    /**
+     * Translates {@link TableSink} into a {@link Transformation}.
+     *
+     * @param withChangeFlag Set to true to emit records with change flags.
+     * @return The {@link Transformation} that corresponds to the translated {@link TableSink}.
+     */
+    protected abstract Transformation<T> translateToTransformation(
```

Comment:
how about we just make the abstract interface which just returns the converted inputRowType, like: createActualSinkInputType. the name and logic is a bit of confusing here.